core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -591,20 +599,25 @@ private[log] class Cleaner(val id: Int,
         currentSegmentOpt = nextSegmentOpt
       }
 
-      cleaned.onBecomeInactiveSegment()
-      // flush new segment to disk before swap
-      cleaned.flush()
-
-      // update the modification date to retain the last modified date of the original files
-      val modified = segments.last.lastModified
-      cleaned.lastModified = modified
-
-      // swap in new segment
-      info(s"Swapping in cleaned segment $cleaned for segment(s) $segments in log $log")
-      log.replaceSegments(List(cleaned), segments)
+      cleanedOpt.foreach(cleaned => {
+        cleaned.onBecomeInactiveSegment()
+        // flush new segment to disk before swap
+        cleaned.flush()
+
+        // update the modification date to retain the last modified date of the original files
+        cleaned.lastModified = segments.last.lastModified
+      })
+
+      if (cleanedOpt.isEmpty) {
+        info(s"All records in segments(s) $segments in log $log are deleted after cleaning. Removing segment(s) $segments.")
+        log.replaceSegments(List(), segments)
+      } else {
+        info(s"Swapping in cleaned segment ${cleanedOpt.get} for segment(s) $segments in log $log")
+        log.replaceSegments(List(cleanedOpt.get), segments)
```

Comment:
Hmm, the changes here are more significant than I expected. I am wondering if they are truly necessary. For example, could we just keep the existing logic and check the size of cleaned segment? If the cleaned segment has size of 0, then we can replace the existing segments with an empty list and delete cleaned.